### ISSUE TITLE
False-positive leak of private type

### ISSUE THREAD

**Issue by derekdreery** [rust-lang/rust#40179]
The following code refuses to compile, although I would expect it to

```rust
#[derive(Debug)]
enum Foo {
    Var1,
    Var2
}

mod bar {
    use super::Foo;

    pub fn get_foo() -> Foo {
        Foo::Var1
    }
}

fn main() {
    println!("{:?}", bar::get_foo());
    println!("Hello, world!");
}
```

The type Foo doesn't leak (it only leaks back into where it was declared), however rustc says it does.

**Comment by petrochenkov** [comment]
This works according to the current rules.
`get_foo` is marked as `pub`, so it's considered potentially visible everywhere, including other crates (even if it's not actually reexported). `Foo` is private and therefore has lower visibility than `get_foo`.

The official workaround looks like this:
```rust
mod detail {
    #[derive(Debug)]
    pub enum Foo {
        Var1,
        Var2
    }
}
use detail::Foo;

mod bar {
    use super::Foo;

    pub fn get_foo() -> Foo {
        Foo::Var1
    }
}

fn main() {
    println!("{:?}", bar::get_foo());
    println!("Hello, world!");
}
```

Another solution is to use "public for the current crate" `pub(crate)` instead of `pub`, it's going to be stabilized [soon](https://github.com/rust-lang/rust/issues/32409):
```rust
#![feature(pub_restricted)]

#[derive(Debug)]
enum Foo { // Foo is `pub(crate)` too if it's defined in the crate root
    Var1,
    Var2
}

mod bar {
    use super::Foo;

    pub(crate) fn get_foo() -> Foo {
        Foo::Var1
    }
}

fn main() {
    println!("{:?}", bar::get_foo());
    println!("Hello, world!");
}
```

**Comment by derekdreery** [comment]
Thanks for the response. :)
